### PR TITLE
imp(evm): stateless custom precompiles

### PR DESCRIPTION
**Pull request by fedekunze**
## Description

This PR refactors the EVM module to support custom stateless precompiled contracts by wrapping the go-ethereum EVM into an interface.

This is the first step toward supporting modular EVM implementations with stateful precompiles. 

**Note**: the concrete implementation and tests will be added in a follow-up PR, since the EVM wrapper is not hooked into the state transition at the moment.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

**Review by Vvaradinov — COMMENTED** [review]
Overall good usage of the geth library for the purpose of precompiled contracts. Fixed a couple of typos in the comments above functions. The only other suggestion I see appropriate is the creation of an additional function that will handle `RunStatefulPrecompiledContract`.

**Review by CoderShiun — COMMENTED** [review]
LGTM.

File: x/evm/statedb/journal.go [hunk 1]
```diff
@@ -24,21 +24,21 @@ import (
...
 
-	// dirtied returns the Ethereum address modified by this journal entry.
-	dirtied() *common.Address
+	// Dirtied returns the Ethereum address modified by this journal entry.
+	Dirtied() *common.Address
 }
 
 // journal contains the list of state modifications applied since the last state
-// commit. These are tracked to be able to be reverted in the case of an execution
+// commit. These are tracked to be able to be Reverted in the case of an execution
```
Comment (emailtovamos) [inline x/evm/statedb/journal.go:38]: No need for capital R. Probably refactoring did it automatically. Same for some other comments

File: x/evm/vm/geth/geth.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+// NewEVM defines the constructor function for the go-ethereum EVM. It uses the
+// default precompiled contracts and the EVM from go-ethereum.
+func NewEVM(
+	blockCtx vm.BlockContext,
+	txCtx vm.TxContext,
+	stateDB vm.StateDB,
+	chainConfig *params.ChainConfig,
+	config vm.Config,
+	_ evm.PrecompiledContracts,
```
Comment (emailtovamos) [inline x/evm/vm/geth/geth.go:31]: Why is this required here? I mean I didn't see the custom contracts used anywhere.

File: x/evm/vm/geth/geth.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+// Config returns the configuration options for the EVM.
+func (e EVM) Config() vm.Config {
+	return e.EVM.Config
+}
+
+// Precompile returns the precompiled contract associated with the given address
+// and the current chain configuration. If the contract cannot be found it returns
+// nil.
+func (e EVM) Precompile(addr common.Address) (p vm.PrecompiledContract, found bool) {
+	precompiles := GetPrecompiles(e.ChainConfig(), e.Context().BlockNumber)
```
Comment (emailtovamos) [inline x/evm/vm/geth/geth.go:57]: Potential for nil pointer error. It should be ensured `e.Context()` is not nil

File: x/evm/vm/interface.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+	Reset(txCtx vm.TxContext, statedb vm.StateDB)
+	Cancel()
+	Cancelled() bool //nolint
+	Interpreter() *vm.EVMInterpreter
+	Call(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	CallCode(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	DelegateCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	StaticCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	Create(caller vm.ContractRef, code []byte, gas uint64, value *big.Int) (ret []byte, contractAddr common.Address, leftOverGas uint64, err error)
+	Create2(caller vm.ContractRef, code []byte, gas uint64, endowment *big.Int, salt *uint256.Int) (ret []byte, contractAddr common.Address, leftOverGas uint64, err error)
```
Comment (emailtovamos) [inline x/evm/vm/interface.go:35]: I am not sure how `Create2` is supposed to be different from Create but either the name should have that info somehow or this function should have a description.

File: x/evm/vm/interface.go [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+// PrecompiledContracts defines a map of address -> precompiled contract
+type PrecompiledContracts map[common.Address]vm.PrecompiledContract
+
+type StatefulPrecompiledContract interface {
+	vm.PrecompiledContract
+	RunStateful(evm EVM, addr common.Address, input []byte, value *big.Int) (ret []byte, err error)
+}
+
+// EVM defines the interface for the Ethereum Virtual Machine used by the EVM module.
+type EVM interface {
```
Comment (emailtovamos) [inline x/evm/vm/interface.go:21]: This seems very general a name. Could you mention a bit more on what this is supposed to do? And its implementation?

File: x/evm/statedb/journal.go [hunk 6]
```diff
@@ -24,21 +24,21 @@ import (
...
 )
 
-// journalEntry is a modification entry in the state change journal that can be
-// reverted on demand.
-type journalEntry interface {
-	// revert undoes the changes introduced by this journal entry.
-	revert(*StateDB)
+// JournalEntry is a modification entry in the state change journal that can be
+// Reverted on demand.
+type JournalEntry interface {
```
Comment (emailtovamos) [inline x/evm/statedb/journal.go:29]: Why did it need exporting? I can't see it used outside in this PR.

File: x/evm/vm/geth/geth.go [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+// ActivePrecompiles returns a list of all the active precompiled contract addresses
+// for the current chain configuration.
+func (EVM) ActivePrecompiles(rules params.Rules) []common.Address {
+	return vm.ActivePrecompiles(rules)
+}
+
+// RunPrecompileContract runs a stateless precompiled contract and ignores the address and
+// value arguments. It uses the RunPrecompiledContract function from the vm package.
+func (EVM) RunPrecompileContract(p evm.StatefulPrecompiledContract, _ common.Address, input []byte, suppliedGas uint64, _ *big.Int) (ret []byte, remainingGas uint64, err error) {
```
Comment (emailtovamos) [inline x/evm/vm/geth/geth.go:70]: Would it make sense to add a test for it?

File: x/evm/vm/interface.go (hunk 4 above)
Comment (4rgon4ut) [reply x/evm/vm/interface.go:35]: @emailtovamos Thanks for your thoughts on this. I must disagree with you, since [CREATE2](https://www.evm.codes/#f5) is well-know EVM op_code using for deploy contracts at a consistent, deterministic address. So, It is not necessary to extra describe it. Also exact name is necessary for EVM interface implementation.

File: x/evm/vm/geth/geth.go (hunk 2 above)
Comment (CoderShiun) [reply x/evm/vm/geth/geth.go:31]: Because type Constructor func() in interface.go
If  _ evm.Constructor = NewEVM, then the variable needs to be there.
For example, NewEVM function can also be like this:

{
var c evm.Constructor
return c(blockCtx, txCtx, stateDB, chainConfig, config, nil)
}

File: x/evm/vm/interface.go [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+	CallCode(caller vm.ContractRef, addr common.Address, input []byte, gas uint64, value *big.Int) (ret []byte, leftOverGas uint64, err error)
+	DelegateCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	StaticCall(caller vm.ContractRef, addr common.Address, input []byte, gas uint64) (ret []byte, leftOverGas uint64, err error)
+	Create(caller vm.ContractRef, code []byte, gas uint64, value *big.Int) (ret []byte, contractAddr common.Address, leftOverGas uint64, err error)
+	Create2(caller vm.ContractRef, code []byte, gas uint64, endowment *big.Int, salt *uint256.Int) (ret []byte, contractAddr common.Address, leftOverGas uint64, err error)
+	ChainConfig() *params.ChainConfig
+
+	ActivePrecompiles(rules params.Rules) []common.Address
+	Precompile(addr common.Address) (vm.PrecompiledContract, bool)
+	RunPrecompileContract(p StatefulPrecompiledContract, addr common.Address, input []byte, suppliedGas uint64, value *big.Int) (ret []byte, remainingGas uint64, err error)
```
[the remark this answers is not in the document]
Comment (4rgon4ut) [reply x/evm/vm/interface.go:40]: nit: Function `RunPrecompileContract` name may be `RunPrecompiledContract` to be more explicit

File: x/evm/vm/geth/geth.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (4rgon4ut) [reply x/evm/vm/geth/geth.go:70]: optional: Function description says that it 
> runs a stateless precompiled contract

but it takes `evm.StatefulPrecompiledContract` as a first argument. I think would be a good idea to add more explanation for that.

**Review by 4rgon4ut — CHANGES_REQUESTED** [review]

File: x/evm/vm/geth/geth.go (hunk 2 above)
Comment (4rgon4ut) [reply x/evm/vm/geth/geth.go:31]: @CoderShiun Thanks for your explanation. Could I kindly bother you to wrap code sections in triple quotes (\```code section\```) for better readability? Also you can use single quotes (\`word\`) for highlighting something like `here`

File: x/evm/statedb/journal.go (hunk 6 above)
Comment (itsdevbear) [reply x/evm/statedb/journal.go:29]: Required for Stateful Precompiles down the line.

File: x/evm/vm/interface.go (hunk 5 above)
Comment (itsdevbear) [reply x/evm/vm/interface.go:21]: 'GethEVM' is what I'd suggest.

File: x/evm/vm/geth/geth.go (hunk 7 above)
Comment (fedekunze) [reply x/evm/vm/geth/geth.go:70]: this PR is just for the wrapper, I'll create a test once the concrete implementation PR is created

File: x/evm/vm/interface.go (hunk 5 above)
Comment (fedekunze) [reply x/evm/vm/interface.go:21]: this is just the EVM interface (common set of funcs). The Geth EVM is one example of a concrete implementation

File: x/evm/vm/interface.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+// PrecompiledContracts defines a map of address -> precompiled contract
+type PrecompiledContracts map[common.Address]vm.PrecompiledContract
+
+type StatefulPrecompiledContract interface {
+	vm.PrecompiledContract
+	RunStateful(evm EVM, addr common.Address, input []byte, value *big.Int) (ret []byte, err error)
+}
+
+// EVM defines the interface for the Ethereum Virtual Machine used by the EVM module.
+type EVM interface {
```
Comment (danburck) [inline x/evm/vm/interface.go:21]: @fedekunze how did you arrive at this list? E.g. `(evm *EVM) Config` is not implemented on geth

**Review by danburck — APPROVED** [review]
@fedekunze LGTM

It would be helpful to add more context to this PR placing it into the grand scheme of the overall spec.

**Comment by loredanacirstea** [comment]
I based my stateful and state-using precompiles work on modifying geth to receive a custom precompile list. 
But I like @fedekunze 's approach better. Wrapping the `go-ethereum/core/vm` constructor into an interface that remains compatible with itself is better and does not require changes from the go-ethereum team.
I wish I knew more about Go interfaces and this trick back in February:
```
type EVM struct {
	*vm.EVM
}
```
I have not tested this PR yet, but I will use this as a base (instead of https://github.com/evmos/ethermint/pull/1131#issuecomment-1247748825), for my precompiles work and see how it works.

**Review by loredanacirstea — APPROVED** [review]
utACK
Wrapping the `go-ethereum/core/vm` constructor into an interface that remains compatible with itself is better and does not require changes from the go-ethereum team.

File: x/evm/vm/interface.go (hunk 9 above)
Comment (fedekunze) [reply x/evm/vm/interface.go:21]: we need to access the public fields from the EVM as well
